backend/pkg/database/refresh_materialized_views.go

```diff
@@ -0,0 +1,163 @@
+}
+
+func (vh *ViewRefreshHandler) GetAndClearChanges() map[string]bool {
+	vh.mu.Lock()
+	defer vh.mu.Unlock()
+	tables := vh.tableChanged
+	vh.tableChanged = make(map[string]bool)
+	return tables
+}
+
+func (vh *ViewRefreshHandler) runRequiredRefreshes(db *gorm.DB) {
+	viewNames := vh.getViewsToRefresh()
+	for viewName := range viewNames {
+		vh.refreshFunc[viewName](db)
+	}
+}
+
+// getViewsToRefresh creates a list of views that should be refreshed due to table changes.
+// The viewsToRefresh is the map of tables that shows which views should be refreshed due to table changes.
+func (vh *ViewRefreshHandler) getViewsToRefresh() map[string]bool {
+	viewToRefresh := make(map[string]bool)
+	tables := vh.GetAndClearChanges()
+	for table := range tables {
+		for _, viewName := range vh.viewsToRefresh[table] {
+			viewToRefresh[viewName] = true
+		}
+	}
+
+	return viewToRefresh
+}
+
+func (vh *ViewRefreshHandler) RegisterViewRefreshHandlers(dbDriver string) {
+	switch dbDriver {
+	case DBDriverTypePostgres:
+		vh.registerViewRefreshHandler(getPostgresViewRefreshHandlerFunc(applicationViewName),
+			applicationViewName,
+			[]string{
+				applicationTableName,
+				resourceTableName,
+				packageTableName,
+				vulnerabilityTableName,
+			})
+		vh.registerViewRefreshHandler(getPostgresViewRefreshHandlerFunc(resourceViewName),
+			resourceViewName,
+			[]string{
+				applicationTableName,
+				resourceTableName,
+				packageTableName,
+				vulnerabilityTableName,
+			})
+		vh.registerViewRefreshHandler(getPostgresViewRefreshHandlerFunc(packageViewName),
+			packageViewName,
+			[]string{
+				applicationTableName,
+				resourceTableName,
+				packageTableName,
+				vulnerabilityTableName,
+			})
+		vh.registerViewRefreshHandler(getPostgresViewRefreshHandlerFunc(vulnerabilityViewName),
+			vulnerabilityViewName,
+			[]string{
+				applicationTableName,
+				resourceTableName,
+				packageTableName,
+				vulnerabilityTableName,
+			})
+	}
+}
+
+func (vh *ViewRefreshHandler) registerViewRefreshHandler(f refreshFunc, viewName string, tables []string) {
+	vh.refreshFunc[viewName] = f
+	vh.addViewsToRefreshByTable(viewName, tables)
+}
+
+func (vh *ViewRefreshHandler) IsSetViewRefreshHandler() bool {
+	return len(vh.refreshFunc) > 0
+}
+
+func (vh *ViewRefreshHandler) addViewsToRefreshByTable(viewName string, tables []string) {
+	for _, table := range tables {
+		vh.viewsToRefresh[table] = append(vh.viewsToRefresh[table], viewName)
+	}
+}
+
+func (db *Handler) SetMaterializedViewHandler(config *DBConfig) {
+	db.ViewRefreshHandler = &ViewRefreshHandler{
+		tableChanged:              make(map[string]bool),
+		viewsToRefresh:            make(map[string][]string),
+		viewRefreshIntervalSecond: time.Duration(config.ViewRefreshIntervalSecond) * time.Second,
+		refreshFunc:               map[string]refreshFunc{},
+	}
+}
+
+func (db *Handler) RefreshMaterializedViews() {
+	if !db.ViewRefreshHandler.IsSetViewRefreshHandler() {
+		return
+	}
```

Comment:
I don't think this should be needed, the view handler should just always be started and running in the background, whether it does anything should be entirely controlled by registering refresh handlers at the point we are creating the views.